presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java

```diff
@@ -0,0 +1,307 @@
+ *     - sourceJoinSymbol: plan B
+ *     - filteringSourceJoinSymbol: symbol b
+ *     - semiJoinOutput: semijoinresult
+ * </pre>
+ */
+public class TransformUncorrelatedInPredicateSubqueryToSemiJoin
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return rewriteWith(new InPredicateRewriter(idAllocator, symbolAllocator), plan, null);
+    }
+
+    /**
+     * Finds InPredicate and calls {@link InsertSemiJoinRewriter} rewriter for each, then replace expressions that
+     * InPredicate is replaced by semi join symbol returned from the used nested rewriter.
+     */
+    private class InPredicateRewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+
+        public InPredicateRewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            node = (FilterNode) context.defaultRewrite(node, context.get());
+
+            List<InPredicate> inPredicates = extractInPredicates(ImmutableList.of(node.getPredicate()));
+            ImmutableMap.Builder<InPredicate, Expression> inPredicateMapping = ImmutableMap.builder();
+            node = rewriteInPredicates(node, inPredicates, inPredicateMapping);
+            return new FilterNode(
+                    node.getId(),
+                    getOnlyElement(node.getSources()),
+                    replaceExpression(node.getPredicate(), inPredicateMapping.build()));
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            node = (ProjectNode) context.defaultRewrite(node, context.get());
+
+            List<InPredicate> inPredicates = extractInPredicates(node.getAssignments().values());
+            ImmutableMap.Builder<InPredicate, Expression> inPredicateMapping = ImmutableMap.builder();
+            node = rewriteInPredicates(node, inPredicates, inPredicateMapping);
+            return new ProjectNode(node.getId(),
+                    getOnlyElement(node.getSources()),
+                    rewriteAssignments(node, inPredicateMapping.build()));
+        }
+
+        private <T extends PlanNode> T rewriteInPredicates(PlanNode node, List<InPredicate> inPredicates, ImmutableMap.Builder<InPredicate, Expression> inPredicateMapping)
+        {
+            for (InPredicate inPredicate : inPredicates) {
+                InsertSemiJoinContext insertSemiJoinContext = new InsertSemiJoinContext(inPredicate);
```

Comment:
I don't think you need the InsertSemiJoinContext class. Just make the InsertSemiJoinRewriter stateful and do something like:

```java
InsertSemijoinRewriter rewriter = new ...;
PlanNode rewritten  = node;
for (InPredicate predicate : inPredicates) {
    rewritten = rewriteWith(rewriter, rewritten);
}

return new NodeWithMappings(rewritten, rewriter.getMappings());
```